Enhance docs

dpordomingo opened the pull request:
caused by https://github.com/src-d/lookout/issues/330
blocks https://github.com/src-d/lookout/pull/398

I did:
- added the `.github/CODE_OF_CONDUCT.md` file,
- moved the example analyzers into `examples` dir,
- fixed the `LICENSE.md` file, and created the `NOTICE.md` file,
- added the `DCO` file,
- added the `CONTRIBUTING.md` file
- added the `make toc` rule into the `Makefile,
- enhanced the `README.md` considering that **source{d} lookout** docs will be in `src-d/lookout` repo,

smacker commented on NOTICE.md:
```diff
@@ -0,0 +1,13 @@
+Copyright (C) 2018 Sourced Technologies, S.L.
```
I don't see this file in our guide. Could you please point why do we need it?

dpordomingo replied:
Not yet merged, but agreed after https://github.com/src-d/guide/pull/305#issuecomment-436596521
It's also added as part of the `project template` by https://github.com/src-d/guide/pull/273 (not yet merged, but almost there)

smacker commented on README.md:
```diff
@@ -1,117 +1,121 @@
+# Caveats
+
+When the gRPC client and servers are being used, it is needed:
+- to set [max gRPC message size](https://github.com/grpc/grpc/issues/7927); using lookout-sdk helpers it can be done:
```
maybe we need to explain why we change defaults. (UAST)

dpordomingo replied:
I agree.

Just to confirm that I understood it, I'd reword as:
>_to set a shared maximum gRPC message size in gRPC servers and clients (to avoid hitting different gRPC limits [see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927)]); the `Analyzer` service should do so when dialing `DataService` and serving `NotifyReviewEvent` and `NotifyPushEvent`; using lookout-sdk helpers it can be done:

smacker commented on README.md:
```diff
@@ -1,117 +1,121 @@
-lookout SDK [![GoDoc](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0?status.svg)](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0/pb) [![PyPI version](https://badge.fury.io/py/lookout-sdk.svg)](https://pypi.org/project/lookout-sdk/) [![Build Status](https://travis-ci.org/src-d/lookout-sdk.svg)](https://travis-ci.org/src-d/lookout-sdk)
------------
+<h1 align="center">
```
according to our CTO html in README should be avoided:
https://github.com/src-d/backlog/issues/1318#issuecomment-435921226

dpordomingo replied:
I also read https://github.com/src-d/guide/pull/292#issuecomment-436597861 
>_Engine has this version because is a product [...]_

So since **source{d} lookout** is also a product, I just coppied the header from **source{d} engine**.

Has @src-d/product any guidelines about product's `README.md`?

smacker replied:
I think we need input from @campoy or @mcuadros to move forward with it.

carlosms replied:
I prefer the html headers, but @mcuadros was quite clear on rejecting its use.
I'd say the safest route would be to keep `lookout-sdk` without html, since it's a lib, and use html only in lookout.

dpordomingo replied:
I understood that **source{d} lookout** would be considered as a whole product, regardless it is separated in different repos, apps, libraries...
But, I'm ok with drop the HTML, that is hard to maintain, and change it again if @src-d/product requires it :dagger:

dpordomingo replied:
I understood that **source{d} lookout** would be considered as a whole product, regardless it is separated in different repos, apps, libraries...
But, I'm ok with drop the HTML, that is hard to maintain, and change it again if @src-d/product requires it :dagger:

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+<!--te-->
+
+
+# What does the SDK provide?
```
```suggestion
# What Does the SDK Provide?
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+
+# What does the SDK provide?
+
+_For the complete documentation of **source{d} Lookout**, please take a look into [http://docs.sourced.tech/lookout](http://docs.sourced.tech/lookout)._
```
```suggestion
_For the complete documentation of **source{d} Lookout**, please take a look at [http://docs.sourced.tech/lookout](http://docs.sourced.tech/lookout)._
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+
+**lookout-sdk** provides:
+
+- **proto [definitions](./proto)**,
```
I'd remove the trailing `,`, it looks strange to me

smacker replied:
I also noticed it but I thought it's just me

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+**lookout-sdk** provides:
+
+- **proto [definitions](./proto)**,
+- pre-generated libraries code for [Golang](./pb) and [Python](./python):
```
```suggestion
- pre-generated libraries for [Golang](./pb) and [Python](./python), offering:
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+
+- **proto [definitions](./proto)**,
+- pre-generated libraries code for [Golang](./pb) and [Python](./python):
+  - an easy **access to the DataService API though a gRPC** service to lookout analyzers, so they won't need to deal with actual Git repositories, UAST extraction, programming language detection, etc,
```
This sounds a bit like the lookout analyzers will be serving that DataService
```suggestion
  - an easy **access to the DataService API though a gRPC service**. Lookout will take care of dealing with Git repositories, UAST extraction, programming language detection, etc
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+- **proto [definitions](./proto)**,
+- pre-generated libraries code for [Golang](./pb) and [Python](./python):
+  - an easy **access to the DataService API though a gRPC** service to lookout analyzers, so they won't need to deal with actual Git repositories, UAST extraction, programming language detection, etc,
+  - low-level **helpers to workaround some protobuf/gRPC caveats**,
```
```suggestion
  - low-level **helpers to work around some protobuf/gRPC caveats**,
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+
+# Caveats
+
+When the gRPC client and servers are being used, it is needed:
```
This sounds like you have an option to use or not gRPC. What about
```
For the gRPC client and server please to follow these requirements:
- set...
- support...
```

dpordomingo replied:
```suggestion
For the gRPC client and server please to follow these requirements:
```

dpordomingo commented:
Thanks for your reviews; I addressed all your suggestions and posted two questions.
(DCO problem now because accepting suggestions; I'll fix it once the PR has been accepted)

smacker commented on examples/language-analyzer.go:
```diff
@@ -6,10 +6,11 @@ import (
 	"io"
 	"os"
 
+	"gopkg.in/src-d/lookout-sdk.v0/pb"
```
I asked in a commet above to describe why we need to change the limit above. I see changes but they don't describe the reason. As I mentioned: the only reason is UAST. For a 1mb file it can be huge.
it means we *need* it only in data server but we *can* do it anywhere.

lookoutd uses an internal helper. That helper sets logging, will set metadata and so on. It also changes the limit if I'm not mistaken because of dummy analyzer. Dummy uses the same helper.

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+
+# Caveats
+
+For the gRPC client and server please to follow these requirements:
```
```suggestion
For the gRPC client and server please follow these requirements:
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+# Caveats
+
+For the gRPC client and server please to follow these requirements:
+- to set a shared maximum gRPC message size in gRPC servers and clients (to avoid hitting different gRPC limits [see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927)]); the `Analyzer` service should do so when dialing `DataService` and serving `NotifyReviewEvent` and `NotifyPushEvent`; using lookout-sdk helpers it can be done:
```
This renders with too many parenthesis and brakets: "servers and clients (to avoid hitting different gRPC limits [see grpc/grpc#7927])"

What about:

```markdown
For the gRPC client and server please to follow these requirements:
- set a common maximum gRPC message size in gRPC servers and clients. This is required to avoid hitting different gRPC limits, see [grpc/grpc#7927](https://github.com/grpc/grpc/issues/7927). To do so use the included helpers in lookout-sdk:
  - go: using `pb.NewServer` and `pb.DialContext`.
  - python: using `lookout.sdk.grpc.create_channel`.
- support [RFC 3986 URI scheme](https://github.com/grpc/grpc-go/issues/1911); lookout-sdk includes helpers for this:
  - go: using `pb.ToGoGrpcAddress` and `pb.Listen`.
  - python: using `lookout.sdk.grpc.to_grpc_address`.
```

carlosms commented on README.md:
```diff
@@ -1,117 +1,121 @@
+    <img src="https://godoc.org/gopkg.in/src-d/lookout-sdk.v0?status.svg"
+         alt="GoDoc">
+  </a>
+  <a href="https://docs.google.com/document/d/1pqz-_SHO5BsJE-aa8o_bAY3r5vR67amnWN8-qZc2UgY">
```
Why the link to the design doc? Did you take this from some template/other project?

I don't really like it. Since it's at the same level as the godoc reference some people may follow it expecting reference material. The design doc was useful for the planning phase, but now we are not keeping it up to date with the latest decisions and features. Also, this particular doc is not even well formatted, I would not link it at the same level as the godoc reference... I would not link it at all.

smacker commented:
sorry for the late comment but I just opened readme once again and I don't see much value in toc. The readme is small enough now that toc becomes just a noise for me.
(no need to take any actions in this PR but if you think the same we can update it later)
